packages/network-controller/src/create-auto-managed-network-client.test.ts

```diff
@@ -0,0 +1,460 @@
+import * as createNetworkClientModule from './create-network-client';
+import {
+  CustomNetworkClientConfiguration,
+  InfuraNetworkClientConfiguration,
+  NetworkClientConfiguration,
+  NetworkClientType,
+} from './types';
+
+/**
+ * An object which instructs the NetworkMock class which JSON-RPC request should
+ * be mocked and how that request should respond.
+ *
+ * @property request - The JSON-RPC request that should be mocked.
+ * @property request.method - The JSON-RPC method.
+ * @property request.params - The JSON-RPC params (optional).
+ * @property response - The JSON-RPC response that the request should return.
+ * @property response.result - Specifies the `result` field of the JSON-RPC
+ * response.
+ * @property response.error - Specifies the `error` field of the JSON-RPC
+ * response.
+ * @property error - The error that should be raised upon making the request.
+ * @property implementation - A function that is called to generate the
+ * response. Should return the same interface as the `response` property above.
+ * @property delay - The amount of time that should pass before the response
+ * returns.
+ * @property httpStatus - The HTTP status that the response should have.
+ * @property discardAfterMatching - Usually a mock specification represents one
+ * invocation of a request; if another request matching the same specification
+ * occurs, then an error will be thrown. If you expect this request to occur an
+ * indeterminate amount of times, you can set this to true and an error will no
+ * longer be thrown.
+ * @property beforeCompleting - Sometimes it is useful to do something after the
+ * request is kicked off but before it ends (or, in terms of a `fetch` promise,
+ * when the promise is initiated but before it is resolved). You can pass an
+ * (async) function for this option to do this.
+ */
+type NetworkMockSpecification = FakeProviderStub & {
+  response: FakeProviderResponse & { httpStatus?: number };
+};
+
+/**
+ * Type guard for determining whether the given value is an error object with a
+ * `message` property, such as an instance of Error.
+ *
+ * Copied from @metamask/utils.
+ *
+ * @param error - The object to check.
+ * @returns True or false, depending on the result.
+ */
+function isErrorWithMessage(error: unknown): error is { message: string } {
+  return typeof error === 'object' && error !== null && 'message' in error;
+}
+/**
+ * Get the error message from an unknown error object. If the error object has
+ * a `message` property, that property is returned. Otherwise, the stringified
+ * error object is returned.
+ *
+ * Copied from @metamask/utils.
+ *
+ * @param error - The error object to get the message from.
+ * @returns The error message.
+ */
+function getErrorMessage(error: unknown): string {
+  return isErrorWithMessage(error) ? error.message : String(error);
+}
+
+/**
+ * Handles mocking JSON-requests sent to the network.
+ */
+class MockedNetwork {
```

Comment:
This may look familiar; there have been a [couple](https://github.com/MetaMask/core/blob/633660f579894692a96350e1a94de8d14c24aef0/packages/network-controller/tests/provider-api-tests/helpers.ts#L125) of [different](https://github.com/MetaMask/metamask-extension/blob/8a93c6cc18c89e441067a6a94fadde4f60e95744/app/scripts/controllers/network/network-controller.test.js#L147) variations on this theme elsewhere. That said, it doesn't do anything by default, so we can keep it pretty simple.